{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5bc77be-bb96-4e0e-bd7b-05dc16c86fdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -U nbdev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f9a4f48d-ede6-4f94-b7a8-b77d42e92a46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pyngrok\n",
      "  Downloading pyngrok-7.0.0.tar.gz (718 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m718.7/718.7 kB\u001b[0m \u001b[31m37.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: PyYAML in /usr/lib/python3/dist-packages (from pyngrok) (5.3.1)\n",
      "Building wheels for collected packages: pyngrok\n",
      "  Building wheel for pyngrok (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pyngrok: filename=pyngrok-7.0.0-py3-none-any.whl size=21129 sha256=50d58a8d216eef67fa6cb3b470cdec924a042c03bd70b15416ba0ab237103a53\n",
      "  Stored in directory: /root/.cache/pip/wheels/a1/27/d7/83363a0fd832c7e4be4a73ae8805f9b5410b40570ae8be7d67\n",
      "Successfully built pyngrok\n",
      "Installing collected packages: pyngrok\n",
      "Successfully installed pyngrok-7.0.0\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install pyngrok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8f928d31-632a-4ee3-878e-b9199bc4d7b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Authtoken saved to configuration file: /root/.ngrok2/ngrok.yml\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "t=2023-09-23T02:55:43+0000 lvl=warn msg=\"ngrok config file found at legacy location, move to XDG location\" xdg_path=/root/.config/ngrok/ngrok.yml legacy_path=/root/.ngrok2/ngrok.yml\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<NgrokTunnel: \"https://3234-172-83-13-4.ngrok-free.app\" -> \"http://localhost:7000\">"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyngrok import ngrok\n",
    "!ngrok config add-authtoken 2TcZn9S1EnAWRznc1F3l07EhsDA_5DKkTteGNdVniRR6JMcAU\n",
    "ngrok_tunnel = ngrok.connect(7000)\n",
    "\n",
    "ngrok_tunnel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b2e4dd8e-177d-472d-9e5f-796870f2a39d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[34mPreparing to preview\u001b[39m\u001b[22m\n",
      "\n",
      "\u001b[32mWatching files for changes\u001b[39m\n",
      "\u001b[32mBrowse at \u001b[39m\u001b[4m\u001b[32mhttp://localhost:7000/\u001b[39m\u001b[24m\n",
      "\u001b[32mGET: /\u001b[39m\n",
      "\u001b[32mGET: /notebooks/am_to_pdb.html\u001b[39m\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\u001b[1mpandoc \u001b[22m\n",
      "  to: html\n",
      "  output-file: setup.html\n",
      "  standalone: true\n",
      "  title-prefix: nbs\n",
      "  section-divs: true\n",
      "  html-math-method: mathjax\n",
      "  wrap: none\n",
      "  default-image-extension: png\n",
      "  css:\n",
      "    - styles.css\n",
      "  toc: true\n",
      "  \n",
      "\u001b[1mmetadata\u001b[22m\n",
      "  document-css: false\n",
      "  link-citations: true\n",
      "  date-format: long\n",
      "  lang: en\n",
      "  theme: cosmo\n",
      "  \n",
      "\u001b[32mGET: /setup.html\u001b[39m\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\u001b[1mpandoc \u001b[22m\n",
      "  to: html\n",
      "  output-file: setup.html\n",
      "  standalone: true\n",
      "  title-prefix: nbs\n",
      "  section-divs: true\n",
      "  html-math-method: mathjax\n",
      "  wrap: none\n",
      "  default-image-extension: png\n",
      "  css:\n",
      "    - styles.css\n",
      "  toc: true\n",
      "  \n",
      "\u001b[1mmetadata\u001b[22m\n",
      "  document-css: false\n",
      "  link-citations: true\n",
      "  date-format: long\n",
      "  lang: en\n",
      "  theme: cosmo\n",
      "  \n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\u001b[1mpandoc \u001b[22m\n",
      "  to: html\n",
      "  output-file: setup.html\n",
      "  standalone: true\n",
      "  title-prefix: nbs\n",
      "  section-divs: true\n",
      "  html-math-method: mathjax\n",
      "  wrap: none\n",
      "  default-image-extension: png\n",
      "  css:\n",
      "    - styles.css\n",
      "  toc: true\n",
      "  \n",
      "\u001b[1mmetadata\u001b[22m\n",
      "  document-css: false\n",
      "  link-citations: true\n",
      "  date-format: long\n",
      "  lang: en\n",
      "  theme: cosmo\n",
      "  \n",
      "\u001b[32mGET: /\u001b[39m\n",
      "\u001b[32mGET: /notebooks/am_to_pdb.html\u001b[39m\n",
      "\u001b[32mGET: /notebooks/am_to_pdb.html\u001b[39m\n",
      "\u001b[32mGET: /notebooks/am_to_pdb.html\u001b[39m\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/quarto.py\", line 297, in _f\n",
      "    try: serve_drv.main(res)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 22, in main\n",
      "    if src.suffix=='.ipynb': exec_nb(src, dst, x)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/nbdev/serve_drv.py\", line 15, in exec_nb\n",
      "    nb = read_nb(src)\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 57, in read_nb\n",
      "    res = dict2nb(_read_json(path, encoding='utf-8'))\n",
      "  File \"/usr/local/lib/python3.9/dist-packages/execnb/nbio.py\", line 16, in _read_json\n",
      "    return loads(Path(self).read_text(encoding=encoding, errors=errors))\n",
      "  File \"/usr/lib/python3.9/json/__init__.py\", line 346, in loads\n",
      "    return _default_decoder.decode(s)\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 337, in decode\n",
      "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n",
      "  File \"/usr/lib/python3.9/json/decoder.py\", line 355, in raw_decode\n",
      "    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\n",
      "json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
      "\u001b[1mpandoc \u001b[22m\n",
      "  to: html\n",
      "  output-file: setup.html\n",
      "  standalone: true\n",
      "  title-prefix: nbs\n",
      "  section-divs: true\n",
      "  html-math-method: mathjax\n",
      "  wrap: none\n",
      "  default-image-extension: png\n",
      "  css:\n",
      "    - styles.css\n",
      "  toc: true\n",
      "  \n",
      "\u001b[1mmetadata\u001b[22m\n",
      "  document-css: false\n",
      "  link-citations: true\n",
      "  date-format: long\n",
      "  lang: en\n",
      "  theme: cosmo\n",
      "  \n",
      "\u001b[32mGET: /setup.html\u001b[39m\n",
      "\u001b[32mGET: /setup.html\u001b[39m\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "!nbdev_preview --port 7000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "102218a6-8643-49c4-9a96-f2d5917bea26",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
